test/Stripe.hs

```diff
@@ -218,3 +249,581 @@ chargeTests =
   where
     token = "foo"
     voucher = "bar"
+
+webhookTests :: TestTree
+webhookTests =
+  testGroup "The Stripe charge web hook"
+  [ testCase "If the signature is missing then the response is Bad Request" $ do
+      db <- memory
+
+      let
+        theSRequest = SRequest jsonRequest checkoutSessionCompleted
+        app = makeApp db
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "missing signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is misformatted then the response is Bad Request" $ do
+      db <- memory
+      let
+        app = makeApp db
+        theRequest = signedRequest "Do you like my signature?"
+        theSRequest = SRequest theRequest checkoutSessionCompleted
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "malformed signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is incorrect then no attempt is made to parse the request body and the response is Bad Request" $ do
+      db <- memory
+      let
+        app = makeApp db
+        theRequest = signedRequest $ stripeSignature (WebhookSecretKey "key") timestamp "Some other body"
+        theSRequest = SRequest theRequest checkoutSessionCompleted
+
+      response <- (flip runSession) app $ srequest theSRequest
+      assertEqual "The body reflects the error" (Just $ Failure "invalid signature") (decode . simpleBody $ response)
+      assertEqual "The response is 400" status400 (simpleStatus response)
+      assertNotRedeemable db voucher fingerprint
+
+  , testCase "If the signature is correct and the body is not JSON then the response is Bad Request" $ do
+      db <- memory
+      let
+        nonJSONBody = "Some other body"
+        app = makeApp db
+        theRequest = signedRequest $ stripeSignature webhookSecret timestamp nonJSONBody
+        theSRequest = SRequest theRequest (LBS.fromStrict nonJSONBody)
+
+      response <- (flip runSession) app $ srequest theSRequest
+
+      -- It should fail but we don't really care what the message is.
+      let (Just (Failure _)) = decode . simpleBody $ response
+      assertEqual "The response is 400" status400 (simpleStatus response)
+
+  , testCase "If the signature is correct then the response is OK" $ do
+      db <- assertOkResponse checkoutSessionCompleted
+      -- It has been paid so we should be allowed to redeem it.
+      assertRedeemable db voucher fingerprint
```

Comment:
Just checking - is returning an error for other event types the behavior we want?  As I understand it, all of these events describe something that has already happened and which PaymentServer can't do anything about.  All it can do is return 200 to have Stripe stop telling it about this event or not-200 to have Stripe repeat the same call at some later point.

I can imagine a not-200 response being useful for logging/metrics purposes.  It might be good for visibility if these unhandled events appear on our monitoring dashboard.  Is that what you're thinking?

Reply:
Logging/metrics wasn't what I was thinking about, but it's a good point.
I want errors to look like errors, especially when it's about something as sensitive as payment.
That's not only true for logging on our side, but also on Stripe's side.

My thinking was: If I get a POST and I return 200 it means I was asked to do something, and I have completed it successfully.
It's a transaction.

But now we have the situation that our service is being asked things it does not know about, so it cannot complete them successfully, so I think it should reply with a codified "what do you want?" (4xx?)

Stripe will do retries on non-200, but only a finite number of times, before giving up.
In any case, returning "OK" to make Stripe shut up seems wrong to me.

The doc could be clearer, but are still helpful IMHO:
https://stripe.com/docs/webhooks/best-practices#events-and-retries